router/proxy/transport.go

```diff
@@ -64,26 +67,55 @@ func (t *transport) setStickyBackend(res *http.Response, originalStickyBackend s
 	}
 }
 
-// always sets the response.Request.URL.Host to the last backend that was
-// connected to.
-func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
+func (t *transport) RoundTrip(ctx context.Context, req *http.Request) (*http.Response, error) {
 	// http.Transport closes the request body on a failed dial, issue #875
-	req.Body = &fakeCloseReadCloser{req.Body}
-	defer req.Body.(*fakeCloseReadCloser).RealClose()
+	fakeBodyCloser := &fakeCloser{req.Body}
+	defer fakeBodyCloser.RealClose()
+
+	cancelc := ctx.Done()
+	reqDone := make(chan struct{})
+	defer close(reqDone)
+
+	req.Body = struct {
+		io.Reader
+		io.Closer
+	}{
+		// This special Reader guards against another subtle race: when the context
+		// is Done() before the httpTransport knows about the request. Details here:
+		// https://go-review.googlesource.com/#/c/2320/5/src/net/http/httputil/reverseproxy.go@120
+		Reader: &runOnFirstRead{
+			Reader: req.Body,
+			fn: func() {
+				go func() {
+					select {
+					case <-cancelc:
+						httpTransport.CancelRequest(req)
+					case <-reqDone:
+					}
+				}()
+			},
+		},
+		Closer: fakeBodyCloser,
+	}
 
 	stickyBackend := t.getStickyBackend(req)
 	backends := t.getOrderedBackends(stickyBackend)
 	for _, backend := range backends {
-		req.URL.Host = backend
-		res, err := httpTransport.RoundTrip(req)
-		if err == nil {
-			t.setStickyBackend(res, stickyBackend)
-			return res, nil
+		select {
+		case <-cancelc:
+			return nil, errRequestCanceled
+		default:
+			req.URL.Host = backend
```

Comment:
I normally use an empty default case in this type of scenario to avoid the indentation (can just move all this code down and outdent.